great_expectations/core/batch_definition.py

```diff
@@ -17,32 +17,36 @@
     )
     from great_expectations.datasource.fluent.interfaces import Batch, DataAsset
 
+PartitionerT = TypeVar("PartitionerT", bound=Optional[Partitioner])
 
-class BatchDefinition(pydantic.BaseModel):
+
+class BatchDefinition(pydantic.GenericModel, Generic[PartitionerT]):
     """Configuration for a batch of data.
 
     References the DataAsset to be used, and any additional parameters needed to fetch the data.
     """
 
     id: Optional[str] = None
     name: str
-    partitioner: Optional[Partitioner] = None
+    partitioner: Optional[PartitionerT] = None
     batching_regex: Optional[re.Pattern] = None
 
     # private attributes that must be set immediately after instantiation
-    _data_asset: DataAsset = pydantic.PrivateAttr()
+    # Note that we're using type Any, but the getter setter ensure the right types.
+    # If we actually specify DataAsset, pydantic errors out.
+    _data_asset: Any = pydantic.PrivateAttr()
 
     @property
-    def data_asset(self) -> DataAsset:
+    def data_asset(self) -> DataAsset[Any, PartitionerT]:
```

Comment:
Is is possible to change the `Any` to `DatasourceT` or does it have to be `Any` for thread reason you have `_data_asset: Any`?